src/Queue.hs

```diff
@@ -16,6 +17,7 @@ enqueue :: a -> Queue a -> Queue a
 enqueue x (Q ins outs) = Q (x : ins) outs
 
 dequeue :: Queue a -> (Maybe a, Queue a)
-dequeue q@(Q [] []) = (Nothing, q)
-dequeue (Q ins (out:outs)) = (Just out, Q ins outs)
-dequeue (Q ins []) = dequeue (Q [] (reverse ins))
+dequeue = \case
+  q@(Q [] []) -> (Nothing, q)
+  Q ins (out:outs) -> (Just out, Q ins outs)
+  Q ins [] -> dequeue (Q [] (reverse ins))
```

Comment:
Very much a nit, but when doing a pattern match on a function with one argument, I think `\case` is cleaner, and also has the nice property of not needed to rewrite the function name on every line.